drivers/adc/ad400x/ad400x.c

```diff
@@ -124,22 +172,104 @@ int32_t ad400x_spi_reg_write(struct ad400x_dev *dev,
 	return ret;
 }
 
+#if !defined(USE_STANDARD_SPI)
+/**
+ * Read samples using spi offload engine
+ * @param dev - The device structure.
+ * @param buf - Buffer to hold the conversion results data
+ * @param samples - number of samples to read
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int32_t ad400x_read_data_offload(struct ad400x_dev *dev,
+					uint32_t *buf,
+					uint16_t samples)
+{
+	struct spi_engine_offload_message msg;
+	uint32_t commands_data[2] = {0xFF, 0xFF};
+	int ret;
+	uint32_t spi_eng_msg_cmds[3] = {
+		CS_LOW,
+		READ(2),
+		CS_HIGH
+	};
+
+	ret = spi_engine_offload_init(dev->spi_desc, dev->offload_init_param);
+	if (ret)
+		return ret;
+
+	msg.commands = spi_eng_msg_cmds;
+	msg.no_commands = NO_OS_ARRAY_SIZE(spi_eng_msg_cmds);
+	msg.rx_addr = buf;
+	msg.commands_data = commands_data;
+
+	ret = spi_engine_offload_transfer(dev->spi_desc, msg, samples * 4);
+	if (ret)
+		return ret;
+
+	if (dev->dcache_invalidate_range)
+		dev->dcache_invalidate_range(msg.rx_addr, samples * 4);
+
+	return ret;
+}
+#endif
+
 /**
- * Read conversion result from device.
+ * Read single conversion result from device.
  * @param dev - The device structure.
  * @param adc_data - The conversion result data
  * @return 0 in case of success, negative error code otherwise.
  */
 int32_t ad400x_spi_single_conversion(struct ad400x_dev *dev,
-				     uint32_t *adc_data)
+				     uint8_t *adc_data)
 {
-	uint32_t buf = 0;
 	int32_t ret;
+	uint16_t bytes_number = 2;
+
+	if (ad400x_device_resol[dev->dev_id ] > 16)
+		bytes_number = 3;
+
+#if defined(USE_STANDARD_SPI)
+	/*
+	 * Trigger a conversion
+	 * No need to explicitly add a delay as the meassuered
+	 * lattency to toggle the gpio and spi clk is arround 3usec
+	 */
+	ret = no_os_gpio_set_value(dev->gpio_cnv, 1);
+	if (ret)
+		return ret;
+
+	ret = no_os_gpio_set_value(dev->gpio_cnv, 0);
+	if (ret)
+		return ret;
+#endif
+	/* SDI must remain high */
+	memset(adc_data, 0xFF, bytes_number);
```

Comment:
I am just now seeing this... it looks it is unsafe... I think this should be reworked (the function template), such that the bytes_number are given as a parameter... 
If someone calls this function with a simple uint8_t variable, instead of an array, we would end-up writing wrong memory locations.